### ISSUE TITLE
[DOCS] GENERIC_EVENT

### ISSUE THREAD

**Issue by reynozeros** [RRZE-HPC/likwid#460]
I have some prblems with some counters on a Zen2 architecture.
In the [PPR for this architacture](https://developer.amd.com/wp-content/resources/55803_B0_PUB_0_91.pdf) I found table 21 on page 182, where a combination of events are recommended, that are not documented in the eventlist afterwards.
In particular I wanted to try the described l2 events (e.g. ```Event[0x43F960]``` )

I now would like to try these events out with the ```GENERIC_EVENT```.
In the [wiki of perfctr](https://github.com/RRZE-HPC/likwid/wiki/likwid-perfctr) it says that the event ```0x437805``` is either specifyable with the full hex code or with ```config=0x05,umask=0x78```.
I didn't get the full hex code specification to work and didn't get, where the ```0x43``` went in the example.
Also putting in the ```0x43``` either in the umask or the config option did not bring me the results that I get when I specify the complete hexcode directly to perf in wrappermode for example.

So did I understand the documentation of the ```GENERIC_EVENT``` wrong or should this be more clear how it works?

**Comment by TomTheBear** [comment]
Hi,

the event `0x43F960` is available under the name `REQUESTS_TO_L2_GRP1_ALL_NO_PF`. It's used by the `L2` group. There is no designated way to find an event based on event config and umask. Probably the simplest is `likwid-perfctr -e | grep 0x60 | grep 0xF9`.

The problem with the `GENERIC_EVENT` is that is partly requires knowledge about how LIKWID handles events internally but you are right, I could make that more clear in the docs. The section is quite short. You are the first external user of this feature that I know of.

The `0x43` are not mentioned in the docs because it is set by LIKWID "on-the-way". The `0x400000` is the "enable counting bit" (on x86) and is used by LIKWID to start and stop events. The `0x30000` specifies the `OsUserMode` (on x86) which tells the register to count in user- and kernel-space. LIKWID by default enables only counting in user-space (unless the `KERNEL` option is set), so if you use `GENERIC_EVENT:PMC0:CONFIG=0x05:UMASK=0x78`, it writes `0x17805` to the register `PMC0` and toggles the "enable counting bit" when needed. This changes when using the `perf_event` backend, there it sets the `count_kernel` flag in the event attributes, the "enable counting bit" is toggled using IOCTL calls on the event file descriptor. As you can see, the details can get complicated and documenting all those is difficult. But I'll try.

**Comment by TomTheBear** [comment]
I tried to document it more clearly. Do you think it is enough like this?

I will try to document also the settings for `CONFIG` and `UMASK` to the architecture-specific pages in the wiki. Then users don't need to search for the vendor documentation that often.

**Comment by reynozeros** [comment]
The documentation is clearer like this, thank you. i just was under the impression, that these are not the same events.
If I measure my stream benchmark with perf:
```perf stat -e r43F960,l3_accesses -a likwid-bench -W N:1610612736B:128 -t stream_avx_fma```
then i get for the L2 w/o PF and for the L3 accesses:
```
Performance counter stats for 'system wide':

   213,303,380,942      r43F960                                                     
   213,474,161,103      l3_accesses

      64.379476537 seconds time elapsed
```
With ```likwid-perfctr -m -C 0-127 -g REQUESTS_TO_L2_GRP1_ALL_NO_PF:PMC0,L3_ACCESS:CPMC0 likwid-bench -W N:1610612736B:128 -t stream_avx_fma```
I get
```
+------------------------------------+---------+----------------+--------------+--------------+--------------+
|                Event               | Counter |       Sum      |      Min     |      Max     |      Avg     |
+------------------------------------+---------+----------------+--------------+--------------+--------------+
|      Runtime (RDTSC) [s] STAT      |   TSC   |      7698.2591 |      60.0966 |      60.2145 |      60.1426 |
| REQUESTS_TO_L2_GRP1_ALL_NO_PF STAT |   PMC0  | 26649916000000 | 192754400000 | 396432200000 | 208202468750 |
|           L3_ACCESS STAT           |  CPMC0  |   210121786000 |            0 |   6573715000 | 1.641576e+09 |
+------------------------------------+---------+----------------+--------------+--------------+--------------+
```

From the number of flops done reported by ```likwid-bench``` I calculated, that 274.877.906.944 cachelines should have been streamed.
So where does the discrepancy from these event counters come from, when they should be the same event?

**Comment by TomTheBear** [comment]
I'm not sure about the `l3_accesses` event. `perf` on my Zen2 system does not provide it. Moreover, the `l3_accesses` from `perf` is probably not `L3_ACCESS:CPMC0` because `L3_ACCESS` is a custom event added by me: https://github.com/RRZE-HPC/likwid/blob/master/src/includes/perfmon_zen2_events.txt#L371-L379 (also not listed in your posted PPR). The `perf` config would be something like `amd_l3/event=0x01,umask=0x80/` but my system returns "not supported" for the setting.

So, your last post should tell me: There is a difference when using `REQUESTS_TO_L2_GRP1_ALL_NO_PF:PMCx` or `GENERIC_EVENT:PMCx:CONFIG=0x60:UMASK=0xf9`?

I saw that there is one thing missing in the docs. For AMD Zen2 core PMUs, for example, the event config is in bits 0-7 and bits 32-35 but for the `GENERIC_EVENT` this is one `CONFIG=0xABB` with `0xA` for bits 32-35 and `0xBB` for 0-7. This is unfortunately architecture-specific.

**Comment by reynozeros** [comment]
My likwid is built with perf support, as our admins want to have it that way. So I have no explanation, where this difference comes from.
likwid (L3_CACHE_REQ is the same event as l3_accesses from perf):
```
+------------------------------------+-----------------------------+----------------+--------------+--------------+--------------+
|                Event               |           Counter           |       Sum      |      Min     |      Max     |      Avg     |
+------------------------------------+-----------------------------+----------------+--------------+--------------+--------------+
|      Runtime (RDTSC) [s] STAT      |             TSC             |      7615.8945 |      59.4457 |      59.6425 |      59.4992 |
| REQUESTS_TO_L2_GRP1_ALL_NO_PF STAT |             PMC0            | 26656381000000 | 192249600000 | 394540900000 | 2.082530e+11 |
|         GENERIC_EVENT STAT         | PMC1:CONFIG=0x60:UMASK=0xf9 | 26660859600000 | 192312400000 | 392949500000 | 208287965625 |
|           L3_ACCESS STAT           |            CPMC0            |   209690546000 |            0 |   6574661000 | 1.638207e+09 |
|          L3_CACHE_REQ STAT         |            CPMC2            |   209690500000 |            0 |   6574661000 | 1.638207e+09 |
+------------------------------------+-----------------------------+----------------+--------------+--------------+--------------+
```

perf:
```
Performance counter stats for 'system wide':

   213,367,987,450      r43F960                                                     
   213,533,027,317      l3_accesses                                                 
   213,367,979,038      l2_request_g1.all_no_prefetch                                   
   213,367,964,679      cpu/event=0x60,umask=0xF9/                                   

      64.739579248 seconds time elapsed
```

Regarding the L3 counter for perf, I read somewhere that the ```amd_l3``` unit may be not available if the kernel version is too old.

**Comment by TomTheBear** [comment]
perf/amd/uncore: Allow F19h user coreid, threadmask, and sliceid specification: https://github.com/torvalds/linux/commit/87a54a1fd525f2af8d82becf583c7e836918cf22

perf/amd/uncore: Allow F17h user threadmask and slicemask specification: https://github.com/torvalds/linux/commit/8170f386f19ca7120393c957d4bfbdc07f964ab6

**Comment by reynozeros** [comment]
Seems like slicemask and threadmask are available for me:
```
$ ls /sys/devices/amd_l3/format/
event  slicemask  threadmask  umask
```
It also seems to be configured right on my system:
```
Performance counter stats for 'system wide':
                                  
   213,502,236,454      l3_accesses                                                 
   213,502,235,709      amd_l3/umask=0xff,event=0x4,slicemask=0xF,threadmask=0xFF/
```
But good to know, I didn't quite get what these bits were for before your comment.
Still, those counters don't wuite match the result I was expecting (274.877.906.944), maybe it doesn't count prefetch requests?

My problem is the L2 counter.
For the size I use for ```likwid-bench``` I calculated from the number of flops being done (reported by itself) that 274.877.906.944 cachelines should be streamed through all caches (including the load before the store).
The ```REQUESTS_TO_L2_GRP1_ALL_NO_PF``` event and also when using ```GENERIC_EVENT:PMC1:CONFIG=0x60:UMASK=0xf9``` the counter seems to overcount by a factor of around 100.
If I use perf with ```cpu/event=0x60,umask=0xF9/``` or other event strings that map on the same one, I get the same count as for the L3 counters and if I include the prefetcher counter for L2 I get roughly the amount that I calculated.
So where does the discrepancy between likwid and perf for the same event come from? (see my previously described results)

And I know, that perf as a wrapper gives slightly different results than likwid in marker mode, as I used it, but the differences are small.

**Comment by TomTheBear** [comment]
Why are you using the FLOP rate to calculate?
```
# LIKWID with perf_event backend
$ likwid-perfctr -C 0,1,2,3 -g REQUESTS_TO_L2_GRP1_ALL_NO_PF:PMC0 -m likwid-bench -W N:1610612736B:4 -t stream_avx_fma
Iterations per thread:	64
Size per thread:	402653184
Data volume (Byte):	103079215104
| REQUESTS_TO_L2_GRP1_ALL_NO_PF |   PMC0  |    404241400 |    403922300 |    403961400 |    404033600 |
| REQUESTS_TO_L2_GRP1_ALL_NO_PF STAT |   PMC0  | 1616158700 | 403922300 | 404241400 | 404039675 |
```
CLs per Thread: `(402653184 Byte * 64)/64 Byte/CL = 402653184 CL`
CLs SUM STAT: `(103079215104 Byte)/64 Byte/CL = 1610612736 CL`

So on my AMD Zen2 system, I don't see any problems here as you can see in the outputs.

Since my `amd_l3` perfmon unit does not have the fix, I can just compare the numbers using the first CPU per L3 segment:
```
$ likwid-perfctr  -C 0,3 -g REQUESTS_TO_L2_GRP1_ALL_NO_PF:PMC0,L3_CACHE_REQ:CPMC0 -m bench/likwid-bench -W N:1610612736B:2 -t stream_avx_fma
| REQUESTS_TO_L2_GRP1_ALL_NO_PF |   PMC0  |    404491800 |    404323700 |
|          L3_CACHE_REQ         |  CPMC0  |    406610800 |    405757800 |
$ likwid-perfctr -C 0,3,6,9,12,15,18,21,24,27,30,33,36,39,42,45 -g REQUESTS_TO_L2_GRP1_ALL_NO_PF:PMC0,L3_CACHE_REQ:CPMC0 -m bench/likwid-bench -W N:1610612736B:16 -t stream_avx_fma
| REQUESTS_TO_L2_GRP1_ALL_NO_PF STAT |   PMC0  | 6463242900 | 403874400 | 404301500 | 4.039527e+08 |
|          L3_CACHE_REQ STAT         |  CPMC0  | 6474310100 | 404446700 | 405504500 | 4.046444e+08 |
```

For a more detailed analysis, I need access to a system with fixed `amd_l3` unit. It might be a change in perf_event, so please send a run with `-V 3` on the command line, then I can compare the configurations.

**Comment by reynozeros** [comment]
Counting the CLs from flops or data volume essentially makes no difference, as these two numbers are correlated. The ```Iterations``` and ```Inner loop executions``` did confuse me though as I first took those for counting the CLs, but the seem to have no connection to it.
I just don't know, if the load associated with the store operation should count or not. The ```data volume``` of the output of ```likwid-bench``` at least does not.
The L3 counters at least don't seem to count them if they count accurately.

Anyways, here is my output with ```-V 3``` and just 5 cores:
[likwid_output.txt](https://github.com/RRZE-HPC/likwid/files/8603584/likwid_output.txt)

The L3 counter seem to correctly count the CLs streamed (when not counting the load before a store operation).
I also include some other counters with some connection to the L2, but they all seem to count wrong.

**Comment by TomTheBear** [comment]
The reason is a bug in the multiplexed event handling with perf_event inside LIKWID. It is present in LIKWID 5.2.0 and was one of the main reasons for the 5.2.1 bugfix release. From the Changelog: **Fix for perf_event multiplexing (important!)** So, please update. I have fought with perf_event multiplexing and did not find a good way to make it work all the time.

Why multiplexing happens is sometimes unclear. One reason could be a running PCP monitoring daemon but also other system features might disturb it (NMI watchdog). But also LIKWID's behavior might cause it. When you run `likwid-perfctr`, it sets up the eventset to check for errors but does not "de-setup" before starting the application. The MarkerAPI inside the applications sets up the eventset again.

**Comment by TomTheBear** [comment]
It depends on the event whether write-allocates (read-for-ownership, RFOs) are counted or not.

`likwid-bench` uses the application view, so has no clue about RFOs. Of course, one could add the output but it requires a separate config option per benchmark to tell how many stores are performed on already loaded data and how many rely on the RFO loading. With hardware features like Intel's SpecI2M or ARM's CL claiming at L2, you cannot reliably tell whether these store requires an RFO load though the whole cache hierarchy or not.

**Comment by reynozeros** [comment]
Well, I installed likwid with spack, thinking I would get the newest version by default. With 5.2.1 it seems to work now. Although the L2 counter still are not producing results I would expect, they at least are kind of in the range now.

**Comment by TomTheBear** [comment]
I added [5.2.1 to spack](https://github.com/spack/spack/blob/develop/var/spack/repos/builtin/packages/likwid/package.py#L25) but releases contain sometimes not the latest versions of the packages.

If you want to discuss your expectations and results, we should use a different communication channel ([Matrix chat for example](https://github.com/RRZE-HPC/likwid#documentation)). If your documentation issue with the GENERIC_EVENT is solved, please close the issue. You can open a new issue if you want to discuss your results in an issue, of course.

**Comment by reynozeros** [comment]
It's fine, I noticed, that the counters have the expected value, when I don't use 5 of the 6 PMC counters, but just 3. Probably a multiplexing issue. The documentation issue is fixed.
